### ISSUE TITLE
pipenv venv not activated before Code Runner executes Python script file

### ISSUE THREAD

**Issue by joeparis** [formulahendry/vscode-code-runner#395]
Windows 10 1803 (17134.407)
Python 3.7.1
pipenv 2018.10.13
VSCode 1.29.1
Code Runner 0.9.5

When running a Python script using Code Runner the virtualenv is not activated before the code is run in the terminal for the first time, causing my code fail. The virtualenv is activated after the script fails and I can then run my code successfully until I close the terminal window. A minimal working example is shown below.

![code-runner](https://user-images.githubusercontent.com/2662123/48879987-32b61700-edc3-11e8-920b-0f5f71df1c3d.gif)

The venv is activated in VSCode and the correct Python interpreter is configured.

Here is the Pipfile for the project:

```
[[source]]
url = "https://pypi.org/simple"
verify_ssl = true
name = "pypi"

[packages]
numpy = "*"

[dev-packages]

[requires]
python_version = "3.7"
```

And here are all the Code Runner settings from my settings.json file:

```
{
    "code-runner.fileDirectoryAsCwd": true,
    "code-runner.respectShebang": false,
    "code-runner.ignoreSelection": true,
}
```

While I can get my code to work this is rather annoying. Does anyone have any ideas what might be causing this?

**Comment by waio1990** [comment]
I have the same issue.

**Comment by joeparis** [comment]
> See : #272

I did see that when searching for a solution and I don't it applies. Code Runner fails even when the proper environment has been set or when

```
"code-runner.executorMap": {
"python":"$pythonPath $fullFileName",
}
```

is set in `settings.json`.  The issue is that Code Runner executes the code being run before VSCode has the chance to activate the appropriate venv.

**Comment by cocoaway** [comment]
HI, have you set the right python envinronment using "Command Palette" --> Python: Select interpreter" ?

I have installed pyenv.

**Comment by waio1990** [comment]
I did set the interpreter, however it doesnt activate my base environment untill after the first run of the code, which i find weird.

**Comment by joeparis** [comment]
@cocoaway Yes, I have set the python environment manually. VSCode also detects the presence of the .venv file and activates the correct environment automatically as was the case in the gif I posted above. However, Code Runner fires off before the activation can happen; i.e., before `.venv\Scripts\activate.ps1` gets called. You can also see this happening in my gif above.

**Comment by HtutLynn** [comment]
I was having the same issue too. I use anaconda and code-runner run the python file with base python interpreter despite selecting the virtual env interpreter.
![test](https://user-images.githubusercontent.com/42746452/49597000-f03b2f80-f9a9-11e8-9a66-b5da30e889f0.gif)

**Comment by RorixRebel** [comment]
Same issue here, runner doesnt seem to update when using pipenv virtual env.
got my env properly selected, i had to override the "run custom command" but it seems to be not ideal.

**Comment by astier** [comment]
I have the same issue with conda-environments.

**Comment by rholland** [comment]
same issue with conda environments

**Comment by ericboxer** [comment]
I've been having the same issue. I've changed the executer map for python to :

```bash
"python": "[ -f 'Pipfile' ] python || pipenv run python",
```

It's checks if the Pipfile exists and runs it in the venv, otherwise it will jsut use the system python. I believe this will only work systems using Bash or something similar. Not sure what the windows counterpart would be.

**Comment by dhdavvie** [comment]
Having the same issue here. I seemed to have fixed it by both making it use the integrated terminal (zsh) and setting the executor map like such:
```
    "code-runner.executorMap": {
        "python": "$pythonPath $fullFileName",
    },
    "code-runner.runInTerminal": true
```

Hope this can help someone.

**Comment by CarlGustavAlbertDwarfsteinYung** [comment]
Can confirm this is still an issue with Visual Studio Code 1.37.1 and Code Runner 0.9.13. Seems like Code Runner runs before Visual Studio has the chance to setup env. Works fine in terminal providing you manually activate it but not in output. 

Following the advice above to change executor to
```
"code-runner.executorMap": {
        "python": "$pythonPath $fullFileName",
},
```
completely fixed it for me.  Windows 10

There should be another option `code-runner.activateEnvironment` or similar that should wait for Visual Studio Code `python.terminal.activateEnvironment` .

**Comment by DGarland1** [comment]
Ctrl+Alt+N doesn't appear to work with CodeRunner and Anaconda.  Works on another computer with just the regular Python 3.6 distro installed.

Running VSCode 1.37.1 (user setup) with CodeRunner 0.9.14

Any idea on how to get this to work using the keyboard shortcut?

**Comment by BruceJi7** [comment]
I tried the solutions above, working with flask, and it doesn't run at all.

**Comment by buckybots** [comment]
Simplest solution to this bug is to just run the second suggested piece 👍 
"code-runner.runInTerminal": true   ,by adding this to the settings.json file in .vscode
python code now runs in Terminal and is able to properly set up environment variables before
running the code

**Comment by pconwell** [comment]
@buckybots , can you specify exactly what you are saying? For example, here is my settings.json file which *does not* work correctly on first run:

```
{
    "python.venvPath": "C:/Users/pconwell/.virtualenvs",
    "python.pythonPath": "C:/Users/pconwell/.virtualenvs/project-6oXFOUjc/Scripts/python.exe"

    "code-runner.executorMap": {
        "python": "$pythonPath $fullFileName",
        }
    "code-runner.runInTerminal": true
}
```

**Comment by buckybots** [comment]
I am not using a venv, but I have the same issue with my environment
variables.  I see you did add the "code-runner.runInTerminal" :true
statement which fixed the issue for me.  It might
be that you are missing some commas between  entries.  Try adding after "
.... .exe",  and also after "....$fullFileName",} ,

**Comment by pconwell** [comment]
@buckybots , thanks for the suggestion, but adding the commas does not fix the issue.

> I am not using a venv

Not to be rude, but if you are not using venv and this issue is related to venv... Your solution - while appreciated - is neither related to nor a solution for the issue being discussed.

**Comment by buckybots** [comment]
Actually, I would say you are being rude by stating that you are not being
rude, but at any rate, I responded, because I am pretty sure I was
responding to the issue, its about the environment variables, not really
about them being set in a venv or not.  I used this thread to identify and
fix my issue.  That being that the environment variables were not being
set.  I am pretty sure this is true weather I was using a venv or not.  I
had environment variables, and their values were not set as described in
the thread, and by using the solution of the thread it fixed my issue.  I
was just trying to help anyone else that had the same problem such as
yourself.  I am sorry that my suggestion of your missing commas did not
help you out.

**Comment by pconwell** [comment]
After some experimenting, I found a simple (if ugly) "solution" (definitely not a real solution) that seems to be working for me on Windows. It seems that code runner initially runs the file with a keyboard input waiting to be read (i.e. `msvcrt.kbhit()` is set to true) and it seems to be trying to write `source C:/Users/user/.virtualenvs/pipenv/Scripts/activate' to terminal but that text is being piped AFTER the file starts execution.

Running the file in any other way (as far as I can tell) starts with `msvcrt.kbhit()` set to false, meaning that the terminal is not waiting for keyboard input.

Therefore, an ugly "solution" is to check if `msvcrt.kbhit()` is true or false before continuing to execute the file. If the terminal is waiting for an input, we can 'trick' it with `input("")` which will close out the open keyboard read.

Putting this at the very top of my file seems to work around the issue (on Windows):
```
# workaround for code runner
import msvcrt
if msvcrt.kbhit():
   input("")
```

I'm not really sure why this works, to be honest. I'm really just guessing here, but I'm wondering if code runner IS executing the activation script but is then keeping the text in buffer and it's getting piped to the terminal twice on accident?

Also, just to clarify, here is my settings.json file:
```
{
    "python.venvPath": "C:/Users/user/.virtualenvs",
    "python.pythonPath": "C:/Users/user/.virtualenvs/pipenv/Scripts/python.exe",
    "code-runner.executorMap": {"python": "$pythonPath -u $fullFileName",},
    "code-runner.runInTerminal": true
 }
```

**Comment by llJll** [comment]
Same issue on mac and using venv.

**Comment by YannickGibson** [comment]
> Having the same issue here. I seemed to have fixed it by both making it use the integrated terminal (zsh) and setting the executor map like such:
> 
> ```
>     "code-runner.executorMap": {
>         "python": "$pythonPath $fullFileName",
>     },
>     "code-runner.runInTerminal": true
> ```
> 
> Hope this can help someone.

This fixes the errors, but its just a work around.
The first time I run code runner after opening VS Code:
   1. new Terminal windows gets opened
   2. python runs directly from selected environment
   3. activates the environment 
That approach kind off doesnt need the third step
Did anybody find a little bit more elegant way of forcing (3.) activation before (2.) first execution?
Thanks!
Below i submit how it looks in terminal.
![image](https://user-images.githubusercontent.com/57909721/78504051-d743b980-776a-11ea-87b4-16ad9a258763.png)

**Comment by AmericanY** [comment]
@YannickGibson  am having the same issue! have you managed to get a solution for it?

**Comment by YannickGibson** [comment]
@AmericanY  No, I'm still using the compromise method, but I am looking forward for somebody that has found a better solution.

**Comment by AmericanY** [comment]
@YannickGibson  am using the following way:

```
"code-runner.executorMap": {
        "python": "~\\Desktop\\Lab\\env\\Scripts\\Activate.ps1 && $pythonPath -u $fullFileName"
    }
```

**Comment by YannickGibson** [comment]
@AmericanY I don't actually think this is a good solution. It just forces the activation of environment everytime you run the script, plus the environment in your executor map is fixed. So every time you want to use different environment you got to change the executor map.
If you press CTRL+ SHIFT + P => "Select Interpreter" => *Select your environment*. Then just "$pythonPath -u $fullFileName" will automatically put the selected environment for you.

**Comment by epfarmer** [comment]
I'm sure this is terrible because I don't know what I am doing, but I have a "solution".

```
    "python.terminal.activateEnvironment": false,
    "code-runner.runInTerminal": true,
    "code-runner.executorMap": {
      "python": "pipenv run python $fileName",
      },
```

This works with activateEnvironment set to true, but after the first run it will give you a warning about how you're trying to run an environment inside an environment.

**Comment by bunzel** [comment]
I was facing the same problem. What worked for me was to activate the environment before launching vscode. 

E.g. if you're working with venv and have in your python project folder a venv named ".env"
1. open a terminal and cd into your python project directory
2. activate your python environment with `source .env/bin/activate`
3. start vscode with `code .`
4. launch python file with Code Runner.

**Comment by sid-am-ahd935** [comment]
> > Having the same issue here. I seemed to have fixed it by both making it use the integrated terminal (zsh) and setting the executor map like such:
> > ```
> >     "code-runner.executorMap": {
> >         "python": "$pythonPath $fullFileName",
> >     },
> >     "code-runner.runInTerminal": true
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Hope this can help someone.
> 
> This fixes the errors, but its just a work around.
> 
> **With this fix the first time I run code runner after opening VS Code:**
> 
> 1. New terminal window is opened
> 2. Python runs from selected environment
> 3. Activates the environment
> 
> That approach kind off doesn't need the third step Did anybody find a little bit more elegant way of forcing "(3.) activation of environment" before "(2.) first execution"? Thanks! Below i submit how it looks in terminal. ![image](https://user-images.githubusercontent.com/57909721/78504108-30abe880-776b-11ea-974c-67f981e94f87.png)


Assuming the python path is still pointing to the scripts folder where the 'activate' file is also present, we can also do this
```
"code-runner.executorMap": {
    "python": "source \"$(dirname '$pythonPath')/activate\" && '$pythonPath' -u $fullFileName",
// OR This    "python": "source \"$(dirname '$pythonPath')/activate\" && python -u $fullFileName",
}
```
I know there is a huge risk in doing this, but as long as the virtual environment creator does not start changing the location of it's script, this configuration should work fine.

Please feel free to give any suggestions or point out any bugs or risks in using this.
Thanks a lot for reading... ;-]

**Comment by mrpbennett** [comment]
I guess there is no true solution for this yet? I have tried most things above and yet to get it to work...back to the terminal i go

**Comment by AmericanY** [comment]
someone reached a solution?

**Comment by soymadip** [comment]
hello, looks like this is a very old problem.

Has this a solution now?

**Comment by jackect** [comment]
@soymadip @AmericanY @mrpbennett @sid-am-ahd935 
I apologize for the interruption. Today, I have resolved the issue and would like to show you how I did it. I found the cause of the problem here:

https://github.com/microsoft/vscode-python/blob/091e121df020ee7eaafd5e310535245477d40d99/src/client/common/terminal/service.ts#L84-L87

My solution involved modifying the source code of "code-runner" located at ~/.vscode/extensions/formulahendry.code-runner-0.12.0/out/src/codeManager.js:394. I made changes to the "executeCommandInTerminal" method as follows.
```python
    executeCommandInTerminal(executor, appendFile = true) {
        return __awaiter(this, void 0, void 0, function* () {
            let isNewTerminal = false;
            if (this._terminal === null) {
                this._terminal = vscode.window.createTerminal("Code");
                isNewTerminal = true;
            }
            this._terminal.show(this._config.get("preserveFocus"));
            this.sendRunEvent(executor, true);
            executor = this.changeExecutorFromCmdToPs(executor);
            let command = yield this.getFinalCommandToRunCodeFile(executor, appendFile);
            command = this.changeFilePathForBashOnWindows(command);
            if (this._config.get("clearPreviousOutput") && !isNewTerminal) {
                yield vscode.commands.executeCommand("workbench.action.terminal.clear");
            }
            if (this._config.get("fileDirectoryAsCwd") && isNewTerminal) {
                const cwd = this.changeFilePathForBashOnWindows(this._cwd);
                command=`cd "${cwd}"\n${command}`;
            }
            if(isNewTerminal){
                yield new Promise((resolve) => setTimeout(resolve, 300));
            }
            this._terminal.sendText(command);
        });
    }
```
 Finally, restart VSCode and used "code-runner" again for execution.🥳

**Comment by tkitka** [comment]
> @soymadip @AmericanY @mrpbennett @sid-am-ahd935 I apologize for the interruption. Today, I have resolved the issue and would like to show you how I did it. I found the cause of the problem here:

Thank you!
This worked well for me. Now venv is activated before the actual code runs.

**Comment by soymadip** [comment]
> @soymadip @AmericanY @mrpbennett @sid-am-ahd935 I apologize for the interruption. Today, I have resolved the issue and would like to show you how I did it. I found the cause of the problem here:
> 
> https://github.com/microsoft/vscode-python/blob/091e121df020ee7eaafd5e310535245477d40d99/src/client/common/terminal/service.ts#L84-L87
> 
> My solution involved modifying the source code of "code-runner" located at ~/.vscode/extensions/formulahendry.code-runner-0.12.0/out/src/codeManager.js:394. I made changes to the "executeCommandInTerminal" method as follows.
> 
> ```python
>     executeCommandInTerminal(executor, appendFile = true) {
>         return __awaiter(this, void 0, void 0, function* () {
>             let isNewTerminal = false;
>             if (this._terminal === null) {
>                 this._terminal = vscode.window.createTerminal("Code");
>                 isNewTerminal = true;
>             }
>             this._terminal.show(this._config.get("preserveFocus"));
>             this.sendRunEvent(executor, true);
>             executor = this.changeExecutorFromCmdToPs(executor);
>             let command = yield this.getFinalCommandToRunCodeFile(executor, appendFile);
>             command = this.changeFilePathForBashOnWindows(command);
>             if (this._config.get("clearPreviousOutput") && !isNewTerminal) {
>                 yield vscode.commands.executeCommand("workbench.action.terminal.clear");
>             }
>             if (this._config.get("fileDirectoryAsCwd") && isNewTerminal) {
>                 const cwd = this.changeFilePathForBashOnWindows(this._cwd);
>                 command=`cd "${cwd}"\n${command}`;
>             }
>             if(isNewTerminal){
>                 yield new Promise((resolve) => setTimeout(resolve, 300));
>             }
>             this._terminal.sendText(command);
>         });
>     }
> ```
> 
> Finally, restart VSCode and used "code-runner" again for execution.🥳

**Comment by soymadip** [comment]
> @soymadip @AmericanY @mrpbennett @sid-am-ahd935 I apologize for the interruption. Today, I have resolved the issue and would like to show you how I did it. I found the cause of the problem here:
> 
> https://github.com/microsoft/vscode-python/blob/091e121df020ee7eaafd5e310535245477d40d99/src/client/common/terminal/service.ts#L84-L87
> 
> My solution involved modifying the source code of "code-runner" located at ~/.vscode/extensions/formulahendry.code-runner-0.12.0/out/src/codeManager.js:394. I made changes to the "executeCommandInTerminal" method as follows.
> 
> ```python
>     executeCommandInTerminal(executor, appendFile = true) {
>         return __awaiter(this, void 0, void 0, function* () {
>             let isNewTerminal = false;
>             if (this._terminal === null) {
>                 this._terminal = vscode.window.createTerminal("Code");
>                 isNewTerminal = true;
>             }
>             this._terminal.show(this._config.get("preserveFocus"));
>             this.sendRunEvent(executor, true);
>             executor = this.changeExecutorFromCmdToPs(executor);
>             let command = yield this.getFinalCommandToRunCodeFile(executor, appendFile);
>             command = this.changeFilePathForBashOnWindows(command);
>             if (this._config.get("clearPreviousOutput") && !isNewTerminal) {
>                 yield vscode.commands.executeCommand("workbench.action.terminal.clear");
>             }
>             if (this._config.get("fileDirectoryAsCwd") && isNewTerminal) {
>                 const cwd = this.changeFilePathForBashOnWindows(this._cwd);
>                 command=`cd "${cwd}"\n${command}`;
>             }
>             if(isNewTerminal){
>                 yield new Promise((resolve) => setTimeout(resolve, 300));
>             }
>             this._terminal.sendText(command);
>         });
>     }
> ```
> 
> Finally, restart VSCode and used "code-runner" again for execution.🥳

Yeah, guess that fixes mine tooo, thanks a lot
